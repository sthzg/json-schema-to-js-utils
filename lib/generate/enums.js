const { createWriteStream } = require('fs');
const {
  DEFAULT_FILE_DOCSTRING,
  DEFAULT_PRETTIER_OPTIONS,
} = require('../constants');
const {
  BuildObjectPathsTransform,
  CompileToTemplateTransform,
  FilterEnumsTransform,
  FilterPropertiesTransform,
  jsonSchemaReadable,
  MemorySinkTransform,
  StdoutWritable,
  PrependToFileTransform,
  PrettierTransform,
} = require('../streams');
const { enumModule } = require('../templates');

module.exports = ({ schema, out }) => {
  const write = out ? createWriteStream(out) : new StdoutWritable();

  jsonSchemaReadable(schema)
    .pipe(new FilterPropertiesTransform())
    .pipe(new FilterEnumsTransform())
    .pipe(new BuildObjectPathsTransform())
    .pipe(new CompileToTemplateTransform(enumModule))
    .pipe(new MemorySinkTransform())
    .pipe(new PrependToFileTransform(DEFAULT_FILE_DOCSTRING))
    .pipe(new PrettierTransform(DEFAULT_PRETTIER_OPTIONS))
    .pipe(write);
};
