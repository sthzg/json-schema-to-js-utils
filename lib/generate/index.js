import generateEnumsModule from './enums';
import generateObjectPathsModule from './paths';
import generateSelectorsModule from './selectors';

export {
  generateEnumsModule,
  generateObjectPathsModule,
  generateSelectorsModule,
};
