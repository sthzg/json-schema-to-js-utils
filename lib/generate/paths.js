const { createWriteStream } = require('fs');
const { pipeline: pipelineFn } = require('stream');
const { promisify } = require('util');
const {
  DEFAULT_FILE_DOCSTRING,
  DEFAULT_PRETTIER_OPTIONS,
} = require('../constants');
const {
  BuildObjectPathsTransform,
  CompileToTemplateTransform,
  StdoutWritable,
  FilterPropertiesTransform,
  jsonSchemaReadable,
  MemorySinkTransform,
  PrependToFileTransform,
  PrettierTransform,
} = require('../streams');
const { objectPathConstant } = require('../templates');

const pipeline = promisify(pipelineFn);

module.exports = ({ schema, out }) => {
  const write = out ? createWriteStream(out) : new StdoutWritable();

  return pipeline(
    jsonSchemaReadable(schema),
    new FilterPropertiesTransform(),
    new BuildObjectPathsTransform(),
    new CompileToTemplateTransform(objectPathConstant),
    new MemorySinkTransform(),
    new PrependToFileTransform(DEFAULT_FILE_DOCSTRING),
    new PrettierTransform(DEFAULT_PRETTIER_OPTIONS),
    write
  );
};
