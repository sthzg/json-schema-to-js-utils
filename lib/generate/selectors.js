const { createWriteStream } = require('fs');
const { pipeline: pipelineFn } = require('stream');
const { promisify } = require('util');
const {
  BuildObjectPathsTransform,
  CompileToTemplateTransform,
  StdoutWritable,
  FilterPropertiesTransform,
  jsonSchemaReadable,
  MemorySinkTransform,
  PrettierTransform,
  PrependToFileTransform,
} = require('../streams');
const {
  DEFAULT_GET_IMPORT,
  DEFAULT_FILE_DOCSTRING,
  DEFAULT_PRETTIER_OPTIONS,
} = require('../constants');
const { selectorFunctionModule } = require('../templates');

const pipeline = promisify(pipelineFn);

/**
 * Generates selector functions for all values inside the JSON schema.
 *
 * Selectors are functions that pick a value from an object in a failsafe way or return undefined if
 * the property is not defined on the input object, e.g.
 *
 *     function byPersonAddressStreet(person) {
 *       return get(person, 'address.street');
 *     }
 */
module.exports = ({ schema, out }) => {
  const write = out ? createWriteStream(out) : new StdoutWritable();

  return pipeline(
    jsonSchemaReadable(schema),
    new FilterPropertiesTransform(),
    new BuildObjectPathsTransform(),
    new CompileToTemplateTransform(selectorFunctionModule),
    new MemorySinkTransform(),
    new PrependToFileTransform(DEFAULT_FILE_DOCSTRING, DEFAULT_GET_IMPORT),
    new PrettierTransform(DEFAULT_PRETTIER_OPTIONS),
    write
  );
};
