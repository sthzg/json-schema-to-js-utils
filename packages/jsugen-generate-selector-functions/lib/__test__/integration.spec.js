import schema from '@sthzg/jsugen-core/lib/testResources/example.schema';
import generate from '../index';

test('generate-selector-functions integration test', () => {
  expect.assertions(1);

  generate({ schema }).subscribe(output => {
    expect(output).toMatchSnapshot();
  });
});
