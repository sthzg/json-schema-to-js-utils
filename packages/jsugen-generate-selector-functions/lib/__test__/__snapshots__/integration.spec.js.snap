// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate-selector-functions integration test 1`] = `
"/**
 * @file This file is auto-generated. Do not modify its contents manually.
 */

import get from 'lodash-es/get';

export function byRootLevelString(property, defaultReturn) {
  return get(property, 'rootLevelString', defaultReturn);
}

export function byRootLevelBool(property, defaultReturn) {
  return get(property, 'rootLevelBool', defaultReturn);
}

export function byRootLevelNumber(property, defaultReturn) {
  return get(property, 'rootLevelNumber', defaultReturn);
}

export function byRootLevelEnumArray(property, defaultReturn) {
  return get(property, 'rootLevelEnumArray', defaultReturn);
}

export function byRootLevelArray(property, defaultReturn) {
  return get(property, 'rootLevelArray', defaultReturn);
}

export function byRootLevelFromDefinitionMemberDefinedInDefinition(
  property,
  defaultReturn,
) {
  return get(
    property,
    'rootLevelFromDefinition.memberDefinedInDefinition',
    defaultReturn,
  );
}

export function byRootLevelFromDefinition(property, defaultReturn) {
  return get(property, 'rootLevelFromDefinition', defaultReturn);
}

export function byLevelOneAnyString(property, defaultReturn) {
  return get(property, 'levelOne.anyString', defaultReturn);
}

export function byLevelOneAnyEnumArray(property, defaultReturn) {
  return get(property, 'levelOne.anyEnumArray', defaultReturn);
}

export function byLevelOneAnyArray(property, defaultReturn) {
  return get(property, 'levelOne.anyArray', defaultReturn);
}

export function byLevelOneAnyBool(property, defaultReturn) {
  return get(property, 'levelOne.anyBool', defaultReturn);
}

export function byLevelOneAnyNumber(property, defaultReturn) {
  return get(property, 'levelOne.anyNumber', defaultReturn);
}

export function byLevelOneLevelTwoFoo(property, defaultReturn) {
  return get(property, 'levelOne.levelTwo.foo', defaultReturn);
}

export function byLevelOneLevelTwo(property, defaultReturn) {
  return get(property, 'levelOne.levelTwo', defaultReturn);
}

export function byLevelOne(property, defaultReturn) {
  return get(property, 'levelOne', defaultReturn);
}
"
`;
